### ISSUE TITLE
Incomplete List of package dependencies 

### ISSUE THREAD

**Issue by johnsamuelwrites** [fish-shell/fish-shell#2062]
fish uses a lot of packages that have not been included in the list of package dependencies. 
After compiling, installing and running fish in a linux container, there are a lot of errors arising out of missing package dependencies like the following error
# fish

Welcome to fish, the friendly interactive shell
Type help for instructions on how to use fish
fish: Unknown command 'hostname'
/usr/local/share/fish/functions/fish_prompt.fish (line 1): hostname|cut -d . -f 1
                                                           ^
in command substitution
    called on line 4 of file /usr/local/share/fish/functions/fish_prompt.fish

Similar error occurs while search man pages.

It will be useful if the presence of required packages or binaries (/bin/hostname) is verified before compiling (or installing)

**Comment by ridiculousfish** [comment]
I guess one first step would be to write a script that scans through all bundled fish functions to extract commands that get used.

**Comment by johnsamuelwrites** [comment]
While building a docker container based on an Ubuntu image, I found that the following dependencies are helpful, to build, install and compile. Need to check with Debian and Fedora based containers
..
 build-essential \
       autoconf \
       python \
       man-db \
       coreutils \
       libncurses5-dev \
       git 

https://github.com/johnsamuelwrites/dockerfiles/blob/3593f69812ad820b782f158dea2b6aef1b6230bc/shell/fish-shell/Dockerfile

**Comment by johnsamuelwrites** [comment]
Here's the list for Fedora https://github.com/johnsamuelwrites/dockerfiles/blob/95464d127034bd2b6a3b9116fadfb6a2311ac9fb/shell/fish-shell/fedora/Dockerfile
...
gcc-c++ \
       make \
       automake \
       autoconf \
       python \
       man-pages \
       man-db \
       hostname \
       coreutils \
       ncurses-devel \
       git 
...
What's worth noting here, I needed to install hostname, As @ridiculousfish pointed, there could be more such packages. But fish worked with these settings without giving any errors

**Comment by treeman** [comment]
Can confirm that with slackware 14.1 I needed to install `man-db` to get `fish_update_completions` to work.

Error for reference:

```
Traceback (most recent call last):
  File "/usr/local/share/fish/tools/create_manpage_completions.py", line 963, in <module>
    file_paths.extend(get_paths_from_manpath())
  File "/usr/local/share/fish/tools/create_manpage_completions.py", line 894, in get_paths_from_manpath
    proc = subprocess.Popen(['manpath'], stdout=subprocess.PIPE)
  File "/usr/lib64/python2.7/subprocess.py", line 711, in __init__
    errread, errwrite)
  File "/usr/lib64/python2.7/subprocess.py", line 1308, in _execute_child
    raise child_exception
OSError: [Errno 2] No such file or directory
```

**Comment by krader1961** [comment]
There are many dependencies which are optional such as python as it is only needed if you intend to use the `fish_config` command. It's not clear those should be listed as hard dependencies. On the other hand something like `hostname` is a show stopper and installing fish should ensure it is available.

**Comment by therealpxc** [comment]
> I guess one first step would be to write a script that scans through all bundled fish functions to extract commands that get used.

Did this ever happen? (I can't really tell because the docker links don't work anymore.) I could really use such a script if there is one. It could potentially do very nice things for packaging fish and oh-my-fish packages using the Nix package manager.

For now I've been using a little script which greps on the output of the (now disabled at build time by default) `__fish_parse` function to try to guess what commands are used in fish scripts. It more or less works, but if I could get a script or small program that systematically identifies external command dependencies of fish scripts I would be extremely happy.

**Comment by zanchey** [comment]
No, it's not really been a priority although we'd certainly be keen to hear from anyone having a go.

**Comment by krader1961** [comment]
I feel compelled to point out that dependencies like `automake`, mentioned in [this comment](https://github.com/fish-shell/fish-shell/issues/2062#issuecomment-102806126), are only needed to build from source.  They are not needed, and should not be included, in the list of dependencies to install the fish binary package. Do we even provide source packages? If we do it's not obvious to me. So it doesn't seem necessary to include such dependencies. It is definitely not the right thing to include them in the fish binary package that 99.99% of users install.

And @ridiculousfish's idea to scan the code for commands that can be automatically turned into dependencies seems like an implausible goal. How would such a tool determine which commands are external rather than functions? If it could make such a determination how would it map the command to a distro package? It seems to me the only practical solution is to manually maintain the list of dependencies for each package. Something that is best crowd-sourced. It seems to me that what we need to do is better document how this information is stored and how to update it.

**Comment by therealpxc** [comment]
Even a tool which presented all possible commands-or-functions would be useful to me. Where I would be using it, I can just conjure up an environment containing all the commands I intend to include and then a tool like this can tell me if any are missing, and where builtins are being used by firing up that environment, putting the binaries on the PATH, and calling `type -q`.

**Comment by zanchey** [comment]
We do provide source packages (as GPG-signed tarballs), but given the mess that is installing from source on any system these are really only for distributors and package management systems to ingest.

**Comment by johnsamuelwrites** [comment]
@krader1961 crowdsourcing is one possible solution. However, this will be very difficult to maintain as a long-term solution.

**Comment by therealpxc** [comment]
I think we have an accurate list of the dependencies in the [nixpkgs source](https://github.com/NixOS/nixpkgs/blob/master/pkgs/shells/fish/default.nix#L106). We have to maintain it as part of the package because we always have complete dependencies for all packages. If our list is incomplete, the package is considered broken.

I've been able to identify potential commands by grepping on the output of the (now disabled) `__fish_parse` function. I think in Nix we can eventually use something like that to get a list of possible commands, then run `type` on them in an environment with fish's runtime deps on the PATH. If there are any that it identifies as functions, we can (maybe?) remove them as external dependencies, and if there are any that it thinks are missing, then we can look for the package that provides that binary and add it to the list.

All we need is a parser that walks the tree ignoring conditionals and instead of running commands adds them to a list, then asks `fish` whether they're internal functions or external commands with `type`. Then those executable names can be used to identify packages according to a particular packaging systme. (We could search on the basis of 'provides' directives in RPM, for example, or use rpmquery on existing files if the executables exist and are on the path).

With those tools we could automatically verify that our list of dependencies is accurate (or at least includes all the required things).

**Comment by ofalk** [comment]
Hey all.

Since this issue was reference on Red Hat Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1478779 , I also came across it (since I used to be the package maintainer for RHEL/Fedora - only stand by these days...). Anyway.

For me this sounds a lot like fish (as distribution package) should be split into a core package and extension/plugin/whatever packages, to make sure we don't install too many deps, just because someone needs the base functionality.

Any other suggestions here? I might have some time to look into packaging again and support Andy (who I think is currently resp. for packaging in RHEL/Fedora).

**Comment by terrycloth** [comment]
I can understand how dependencies on some of the utilities that are almost always included in a POSIX system, like maybe `man-db`, were missed when the fish functions were first written. But I'm surprised getting an accurate list of dependencies hasn't been considered more of a priority...

I'm not positive I got this right, but I figured I'd give it a try. Let me know if my grep-fu is weak, or if you notice where I might have missed something. Here's what I did:

``` fish
### Get a list of all files from the installed binary fish package (on Fedora 26 Workstation x86_64).
sudo dnf repoquery --list fish

### ...That produced a lot of non-fish files, so let's narrow down the list, and save it.
sudo dnf repoquery --list fish | grep -i ".*\.fish"  > fish-files.txt

### Let's grep all those files, to make sure we have a comprehensive list of every command ever.
grep --extended-regexp --only-matching '^[[:space:]]*[[:alnum:].+_-]+'  --no-filename ( cat fish-files.txt )  > fish-all-commands.txt

### Okay, now let's trim the leading whitespace, and remove all duplicates.
sed --regexp-extended 's/^[[:space:]]+//g'  fish-all-commands.txt  | sort --unique  > fish-unique-commands.txt
```

Resulting in this list of 362-item list: **[fish-unique-commands.txt](https://github.com/fish-shell/fish-shell/files/1203890/fish-unique-commands.txt)**

None of those include `bc` though, and I remember running into issues when not having that installed once before — was `bc` removed from fish, or is it used "under the hood" by something else?

I tried running the final filter with `type` anyway:

``` fish
type --path ( cat fish-unique-commands.txt ) ^&1 | grep --extended-regexp --invert-match '^type: '  > fish-possible-dependencies.txt
```

I figure it should be easy enough for different distro package maintainers to find the list of packages they need based on this resulting list of executables. Right? **[fish-possible-dependencies.txt](https://github.com/fish-shell/fish-shell/files/1203896/fish-possible-dependencies.txt)**

**Comment by faho** [comment]
@terrycloth: Your approach has a number of problems.

- You're checking completions as well. These often call the tool to be completed, which isn't something I'd call a "dependency". This also applies to functions that are only used in completions.

- It's missing things that are only ever called in command substitutions - this is why `bc` is missing (it's used to implement the `math` function, so it's a fairly important dependency).

- It's misinterpeting continued lines, which is why it includes "UserKnownHostsFile".

- It can't figure out what is used conditionally (e.g. if it exists).

--------

Currently, fish absolutely depends on just a few things:

- make/autoconf/a compiler to build (these don't need to be on the end user's system unless thats the one compiling)

- diff and expect for tests (if you don't want to run our tests, you don't need these)

- ncurses, coreutils (at least ls, seq, rm, mktemp) and bc (and pcre2, though we also include our own copy I'd recommend a system one) for general usage.

Then there's the long tail of stuff we use if available and will (for the most part - if we don't this is a bug) gracefully degrade without (this might include an error message if you manually try to use something that requires it). From memory, this is:

- gettext for localization

- python (for fish_config and fish_update_completions) - this can be, IIRC at a minimum python 2.6 or 3.3.

- man/apropos for generating completions and descriptions (and nroff for printing help) - this should now work with both man-db and mandoc IIRC. Also gunzip if your man pages are gzipped.

- A command called `hostname` for showing the hostname in the prompt (we might add more ways to get the hostname here)

- git/hg/svn for showing vcs information in the prompt

Of these, I'd not add git/hg/svn as dependencies to the package - if you're not using git, you don't need git information in the prompt, and if you are, you already have it. python and the man stuff is nice to have but can be done without, so if you're making a package for a "minimal" OS, this can be "recommended" or "optional".

**Comment by krader1961** [comment]
@faho's analysis is spot on. I honestly don't see the necessity of writing a tool that uses the fish parser to find every potential external command referenced in all the non-completion scripts. The list of essential commands is very small and hasn't changed in the two years since this was opened. And identifying potential external commands is just half the battle. You still need a  distro specific tool to convert that list to package names. And you still have to decide which packages are required for fish to work at all (coreutils, hostname) versus optional to provide an improved experience (man-db, gettext, python). Obviously some distros might consider, for example, python a mandatory dependency since they might consider a working `fish_config` command an essential feature. But I don't think that's something for the fish team to decide.

To sum up, the fish team should maintain two lists: required commands and nice to have but optional commands. Each distro can then decide which, if any, of the optional commands they want to include in the required list and map each list to the appropriate packages for their distro. The question is how should those lists be made available to the people who maintain distro packages?

**Comment by faho** [comment]
One more dependency: xsel or pbcopy/pbpaste for the fish_clipboard_ functions (bound to \cv and \cx). Again strictly optional - if you don't have them, the bindings just won't work.

**Comment by zanchey** [comment]
> To sum up, the fish team should maintain two lists: required commands and nice to have but optional commands. Each distro can then decide which, if any, of the optional commands they want to include in the required list and map each list to the appropriate packages for their distro.

This is the README. If a package that is used to provide optional functionality does not appear in the README, then that should be corrected. If a package listed as optional causes the shell to spew errors if it does not exist (manpage completion, I'm looking at you), then that's usually a bug as well.

**Comment by ofalk** [comment]
OK. I'll leave this discussion to you guys, since there's obviously no consensus about how to best maintain this... If there's a decision, feel free to update RH BZ to let us know if we should change anything regarding packaging on Fedora/RHEL.

**Comment by terrycloth** [comment]
@faho 
> You're checking completions as well. These often call the tool to be completed, which isn't something I'd call a "dependency". This also applies to functions that are only used in completions.

Okay, I can remove the completions directory from the list of files to search. I don't know of an easy way to find functions that are only used *in* completions if they're saved outside the completions directory, but any fish functions themselves get filtered out later in the scan.

> It's missing things that are only ever called in command substitutions...

Ah, yes. Another grep for the first word inside parentheses, and after a pipe, should get those, right?

``` fish
grep --extended-regexp --only-matching '\([[:space:]]*[[:alnum:].+_-]+'  --no-filename ( cat fish-files.txt )
grep --extended-regexp --only-matching '\|[[:space:]]*[[:alnum:].+_-]+'  --no-filename ( cat fish-files.txt )
```

Then the sed command that trims leading whitespace will also need to trim the leading parenthesis and pipe characters.

> It's misinterpeting continued lines, which is why it includes "UserKnownHostsFile".

I think that should be filtered out by the `type` command. "UserKnownHostsFile" appears in the so-called "commands" list file, but not in the final list of external executables.

> It can't figure out what is used conditionally (e.g. if it exists).

True, my scans don't differentiate between hard and optional dependencies. Probably we can be more confident in the mandatory dependencies that you listed and which are already documented, though, as such packages always produce errors when missing. So perhaps this scan is mostly useful for tracking down the last of the lesser-known optional dependencies?

@krader1961
>I honestly don't see the necessity of writing a tool that uses the fish parser to find every potential external command referenced in all the non-completion scripts...

I see this as a one-time effort to help update a specific part of the documentation — optional features and dependencies, at least. Then, hopefully, the docs will keep up to date when any new features and external command calls are added, optional or not. If there were another reason all this wasn't necessary, though, I'd be happy to hear it 😄

> You still need a distro specific tool to convert that list to package names.

Sure, but I figure that's more the responsibility of the distro packagers anyway. (Also, I believe `dnf` and `apt` and the others can look up packages by executable file path.) For the fish documentation, it should be enough to provide recognizable names for any dependencies, with a brief note on what they're used for when the dependency is for an optional feature.

***

To summarize, here's the current draft of my soft-dependency-scanning script:

**[fish-dependency-scanner.txt](https://github.com/fish-shell/fish-shell/files/1210978/fish-dependency-scanner.txt)** (GitHub won't allow uploading a file with a .fish file extension.)

And the new list of executables that it produces:

**[fish-unique-executables.txt](https://github.com/fish-shell/fish-shell/files/1210980/fish-unique-executables.txt)** (down to 64 items)

Once we agree on the list of executables, we can look up which external applications/projects they come from, and we can then document their role and dependency status too.

**Comment by terrycloth** [comment]
...One weakness of the scanner script that just occurred to me: filtering with `type --path` will only work if I happen to have the executable installed already on the machine where I do the testing. My Fedora desktop has fish and lots of other packages installed, but if I do not coincidentally have everything else already, there may be more potential dependencies that I'm not seeing.

**Comment by terrycloth** [comment]
I decided to just not filter out the "could not find" lines from `type`. I guess I'll have to sort out of the false positives myself.

**[fish-dependency-scanner.txt](https://github.com/fish-shell/fish-shell/files/1217049/fish-dependency-scanner.txt)**

**[fish-unique-executables.txt](https://github.com/fish-shell/fish-shell/files/1217050/fish-unique-executables.txt)** (jumped up 318 items 😕)

I guess I'll start looking these up over the next few days...

**Comment by faho** [comment]
>fish-unique-executables.txt (jumped up 318 items 😕)

So, going through that list, I removed all that are not actually called as commands (including basically everything that has an uppercase letter in it, given that this is unix and we do not use our shouty key), everything that is a (non-wrapping) function or builtin (or keyword).

Then I tried to figure out which is only used in a completion script for itself and removed those. Then I removed the ones we already mentioned, and the ones we've removed since the release @terrycloth is checking (this includes `rev`). Note that it's very well possible that I might have missed something.

This is the list:

awk
basename
cat
cut
date
dircolors
dirname
find
getent
kill
mkdir
mkfifo
ps
sort
stat
stty
tail
tput
tr
tty
uname
uniq
wc
whoami

The corresponding archlinux packages are

coreutils
findutils
gawk
glibc
ncurses
procps-ng
util-linux

and this should be reasonably close in other OSs - `kill` is in util-linux here, "gawk" might be "mawk" or just "awk" instead (I think we're reasonably compatible in our use). I would consider these to be essential on any unix system - Arch has them in the "base" group, which is assumed to always be installed (and hence doesn't need to be listed as an explicit dependency on anything).

Do we need to mention these?

**Comment by therealpxc** [comment]
> I  would consider these to be essential on any unix system - Arch has them  in the "base" group, which is assumed to always be installed (and hence  doesn't need to be listed as an explicit dependency on anything).
> 
> Do we need to mention these?

With functional package management systems, all packages are isolated from one another so that they cannot find any libraries but the ones they were built against, and executable commands are, as far as possible, kept off the PATH but instead hardcoded as full paths at compile time. This means that they are independent of one another as well as the base system. It may seem odd, but this results in additional robustness because it allows packages depending on conflicting versions of the same dependencies (even executables with the same basename) to coexist on a system. For functional package management-based Linux distros like NixOS and GuixSD, this means that users can easily install different versions of the same package without affecting one another, while still transparently sharing resources when their needs are the same. When using Nix and Guix as package managers atop other distros, this means that installing alternate versions of libraries and executables with Nix or Guix can (hopefully) never break the functionality of the packages which are part of the underlying distro or vice-versa. Every 'impurity' is an opportunity for such breakage, and to be avoided where possible.

One of the consequences is that for people developing package recipes for functional package managers need a genuinely complete list of all the dependencies, with no presumptions about which packages are always expected as part of the 'base system' (a notion that may vary per-distro).

In the case of Nix, we actually still need to add an explicit dependency on `awk` for completions-- that we haven't done so is an impurity and potential source of bugs. Guix, on the other hand, currently lacks a `fish` package entirely, and a genuinely complete dependency specification would be of interest to anyone working to package `fish` for Guix.

Moreover, since some of those dependencies (`procps` and `utillinux`) are platform-specific, it would be a convenience to packagers for cross-platform package managers (like Nix, which works on Linux, macOS, and FreeBSD, among other platforms) to have them named along with their corresponding platforms, although of course in this case the fact that they are Linux-specific is fairly obvious from their names alone.

Perhaps they can be mentioned in a separate section, or they can be named along with a note which indicates that they are probably already present on most systems, although it is up to the norms of each Linux distro to determine how complete its packages' dependency specifications must be, but I think they should be named in the README.

**Comment by faho** [comment]
>Moreover, since some of those dependencies (procps and utillinux) are platform-specific

The commands we use from them aren't. Like I said, "util-linux" is used here for `kill`, "procps-ng" is used for `ps`. I assume our scripts work with any reasonable version of that.

Which is why in this case I'd rather name the commands than the package.

**Comment by zanchey** [comment]
I made some further changes to the README to improve readability; I think we're done here.

**Comment by terrycloth** [comment]
I thought it made sense to put the build and runtime dependencies together, since, if you install from a package, it should either bundle or use the package manager to pull in _all_ dependencies. The dependencies list is therefore only useful to package maintainers, and to users who want to build from source.

**Comment by zanchey** [comment]
My thinking is that the README is also the landing page for our GitHub project, and as such should detail what users need to know before downloading and installing fish. In particular, what else is it going to drag in on my system, and what might not be available by default (optional dependencies)?
